### PR TITLE
Add DbFunctionParameter support for DbFunctions

### PR DESCRIPTION
**Pull request by pmiddleton**
Adds the ability to map properties for DbFunction parameters.  See #13752

### CODE REVIEW

**Comment by smitpatel** [comment]
@ajcvickers - Should DbFunction's parameter have ValueConverter also?
One weird issue which can come in the absence of it, is if you are passing a column vs a constant. They both would have different types.

**Comment by pmiddleton** [comment]
I need to review this PR this weekend to make sure I didn't miss anything.  

I was not planning implementing the code which is going to use the new properties.  Since the query pipeline is being rewritten it doesn't make sense to update the old one.

@smitpatel - Does the new query work live in the smit/query branch?

**Comment by smitpatel** [comment]
@pmiddleton - Yes, that is the branch. I was thinking the same that we can commit & merge API to master and new query pipeline will make use of it when rebased. It would be good if we have tests to verify the behavior but skipped in old pipeline.

**Comment by ajcvickers** [comment]
@smitpatel I think it needs to be possible to specify the actual type mapping to use (which could have a value converter on it) rather than only allowing the database type name. The database type name could be nice sugar for getting to the type mapping in common cases.

**Comment by smitpatel** [comment]
@pmiddleton - I discussed with @ajcvickers about ValueConverters area. Essentially query needs TypeMapping for parameters so that it can generate literal/parameter correctly in SQL. So IDbFunctionParameter should store RelationalTypeMapping (& also IDbFunction should store RelationalTypeMapping as return type).

Currently, there is no way to specify type mapping directly in fluent API. So for now we can have sugar method to specify database type and run a convention to assign type mapping whenever database type is set. TypeMappingConvention does the same for columns.

**Comment by pmiddleton** [comment]
@smitpatel 

Does it make more sense to look up the mappings at query translation time vs storing it?  I am not completely familiar with the RelationalTypeMapping but is is possible for the mappings to be changed at runtime after model creation?  If so then we shouldn't be precomputing and storing them.

**Comment by smitpatel** [comment]
@pmiddleton - I believe we can differ that for now since IProperty also does not change TypeMapping at runtime.

**Comment by pmiddleton** [comment]
@smitpatel - Ok I will cache the values for parameters and return type, we can change to a realtime lookup later if needed.

**Comment by pmiddleton** [comment]
@smitpatel - Take a look now and let me know if this is what you had in mind for RelationalTypeMapping.  If so I still need to add unit tests.

**Comment by smitpatel** [comment]
Requesting review from @ajcvickers for what to store in function parameter for type mapping.

File: src/EFCore.Relational/Metadata/Internal/DbFunctionParameter.cs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        ///     This API supports the Entity Framework Core infrastructure and is not intended to be used 
+        ///     directly from your code. This API may change or be removed in future releases.
+        /// </summary>
+        public virtual bool SupportsNullabilityPropagation { get; set; }
+
+        /// <summary>
+        ///     This API supports the Entity Framework Core infrastructure and is not intended to be used 
+        ///     directly from your code. This API may change or be removed in future releases.
+        /// </summary>
+        public virtual RelationalTypeMapping TypeMapping { get; set; }
```
Comment (ajcvickers) [inline src/EFCore.Relational/Metadata/Internal/DbFunctionParameter.cs]: @smitpatel Looks like it contains the type mapping, which is great!

File: src/EFCore.Relational/Metadata/Internal/DbFunctionParameter.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        ///     This API supports the Entity Framework Core infrastructure and is not intended to be used 
+        ///     directly from your code. This API may change or be removed in future releases.
+        /// </summary>
+        public virtual string Name { get; set; }
+
+        /// <summary>
+        ///     This API supports the Entity Framework Core infrastructure and is not intended to be used 
+        ///     directly from your code. This API may change or be removed in future releases.
+        /// </summary>
+        public virtual string DatabaseType { get; set; }
```
Comment (ajcvickers) [inline src/EFCore.Relational/Metadata/Internal/DbFunctionParameter.cs]: Assuming this is the store type name, it probably doesn't need to be here in addition to the type mapping.

Comment (pmiddleton) [reply src/EFCore.Relational/Metadata/Internal/DbFunctionParameter.cs]: Removed

**Comment by pmiddleton** [comment]
@smitpatel @ajcvickers - Rebased on latest master.  Was there something else that needed changing for this PR before merging?  

Sorry I'm in process of trying to buy a house and move this spring so I haven't had as much time as I would like to work on things.

**Comment by ajcvickers** [comment]
@smitpatel Are we still waiting for anything here?

File: src/EFCore/EFCore.csproj [hunk 3]
```diff
@@ -1,11 +1,12 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <Description>Entity Framework Core is a lightweight and extensible version of the popular Entity Framework data access technology.
+    <Description>
```
Comment (smitpatel) [inline src/EFCore/EFCore.csproj]: Seems like unrelated change

**Comment by smitpatel** [comment]
>                     if (RelationalDependencies.TypeMappingSource.FindMapping(methodInfo.ReturnType) == null)

This should be outside of Translation == null check. Even if they are providing translation, they cannot pass in TypeMapping for the translated expression.
Also see below for parameter

---
Refers to: src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs:84 in a99735f. [](commit_id = a99735f850e3a45bf6ab8a88f2c978e382b82bb3, deletion_comment = False)

**Comment by smitpatel** [comment]
>                         if (RelationalDependencies.TypeMappingSource.FindMapping(parameter.ParameterType) == null)

This should go away. All parameters on method info should be inside dbFunction.Parameters check below. And that loop should handle parameters

---
Refers to: src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs:94 in a99735f. [](commit_id = a99735f850e3a45bf6ab8a88f2c978e382b82bb3, deletion_comment = False)

File: src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs [hunk 4]
```diff
@@ -100,6 +100,18 @@ protected virtual void ValidateDbFunctions([NotNull] IModel model, DiagnosticsLo
                                     parameter.ParameterType.ShortDisplayName()));
                         }
                     }
+
+                    foreach (var parameter in dbFunction.Parameters)
+                    {
+                        if(RelationalDependencies.TypeMappingSource.FindMapping(parameter.FunctionType) == null)
```
Comment (smitpatel) [inline src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs]: We should first check if a typemapping is specified which can be mapped if not then only we should go and find mapping based on ClrType. That is applicable for both parameters and return value of actual function

File: src/EFCore.Relational/Metadata/IDbFunction.cs [hunk 5]
```diff
@@ -13,8 +14,7 @@ namespace Microsoft.EntityFrameworkCore.Metadata
     /// </summary>
     public interface IDbFunction
     {
-        /// <summary>
-        ///     The name of the function in the database.
+        /// <summary>,abase.
```
Comment (smitpatel) [inline src/EFCore.Relational/Metadata/IDbFunction.cs]: Typo

File: src/EFCore.Relational/Metadata/IDbFunctionParameter.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        ///     This API supports the Entity Framework Core infrastructure and is not intended to be used 
+        ///     directly from your code. This API may change or be removed in future releases.
+        /// </summary>
+        string Name { get; }
+
+        /// <summary>
+        ///     This API supports the Entity Framework Core infrastructure and is not intended to be used 
+        ///     directly from your code. This API may change or be removed in future releases.
+        /// </summary>
+        Type FunctionType { get; }
```
Comment (smitpatel) [inline src/EFCore.Relational/Metadata/IDbFunctionParameter.cs]: Should just call this ClrType

File: src/EFCore.Relational/Metadata/IMutableDbFunctionParameter.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.EntityFrameworkCore.Storage;
+
+namespace Microsoft.EntityFrameworkCore.Metadata
+{
+    /// <summary>
+    ///     Represents a db function parametere in an <see cref="IDbFunction" />.
+    /// </summary>
+    public interface IMutableDbFunctionParameter : IDbFunctionParameter
```
Comment (smitpatel) [inline src/EFCore.Relational/Metadata/IMutableDbFunctionParameter.cs]: I am not sure if we also need conventional derived classes @AndriySvyryd

File: src/EFCore.Relational/Metadata/Builders/DbFunctionParameterBuilder.cs [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        /// </summary>
+        public DbFunctionParameterBuilder(DbFunctionParameter dbFunctionParameter)
+        {
+            _builder = new InternalDbFunctionParameterBuilder(dbFunctionParameter);
+        }
+
+        /// <summary>
+        ///     Sets the name of the parameter
+        /// </summary>
+        public virtual DbFunctionParameterBuilder HasName([NotNull] string name)
```
Comment (smitpatel) [inline src/EFCore.Relational/Metadata/Builders/DbFunctionParameterBuilder.cs]: Can parameter name be mutated? I would assume it would be same as parameter name you specify on your function.

File: src/EFCore.Relational/Metadata/Conventions/Infrastructure/RelationalConventionSetBuilder.cs [hunk 9]
```diff
@@ -74,6 +74,8 @@ public override ConventionSet CreateConventionSet()
             conventionSet.PropertyAddedConventions.Add(relationalColumnAttributeConvention);
 
             var sharedTableConvention = new SharedTableConvention(logger);
+            var dbFunctionTypeMappingConvention = new DbFunctionTypeMappingConvention(RelationalDependencies.TypeMappingSource);
```
Comment (smitpatel) [inline src/EFCore.Relational/Metadata/Conventions/Infrastructure/RelationalConventionSetBuilder.cs]: Initialize closer to usage

File: src/EFCore.Relational/Metadata/Conventions/Internal/DbFunctionTypeMappingConvention.cs [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+            _typeMappingSource = typeMappingSource;
+        }
+
+        /// <summary>
+        ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
+        ///     directly from your code. This API may change or be removed in future releases.
+        /// </summary>
+        public virtual InternalModelBuilder Apply(InternalModelBuilder modelBuilder)
+        {
+            foreach (var dbFunc in modelBuilder.Metadata.Relational().DbFunctions.Cast<IMutableDbFunction>())
```
Comment (smitpatel) [inline src/EFCore.Relational/Metadata/Conventions/Internal/DbFunctionTypeMappingConvention.cs]: nit:Prefer full name like dbFunction rather than Func

File: src/EFCore.Relational/Metadata/Internal/DbFunction.cs [hunk 11]
```diff
@@ -81,7 +84,7 @@ public class DbFunction : IMutableDbFunction, IMethodCallTranslator
                 && !typeof(DbContext).IsAssignableFrom(methodInfo.DeclaringType))
             {
                 throw new ArgumentException(
-                    RelationalStrings.DbFunctionInvalidInstanceType(methodInfo.DisplayName(), methodInfo.DeclaringType.ShortDisplayName()));
+                    RelationalStrings.DbFunctionInvalidInstanceType(methodInfo.DisplayName(), methodInfo.DeclaringType?.ShortDisplayName()));
```
Comment (smitpatel) [inline src/EFCore.Relational/Metadata/Internal/DbFunction.cs]: When would declaring type would be null?

File: src/EFCore.Relational/Metadata/Internal/DbFunctionParameter.cs (hunk 2 above)
Comment (smitpatel) [reply src/EFCore.Relational/Metadata/Internal/DbFunctionParameter.cs]: It is still appearing for me in codeflow review.

---
In reply to: [272869029](https://github.com/aspnet/EntityFrameworkCore/pull/14990#discussion_r272869029) [](ancestors = 272869029)

File: src/EFCore.Relational/Metadata/Builders/DbFunctionBuilder.cs [hunk 12]
```diff
@@ -92,6 +95,29 @@ public virtual DbFunctionBuilder HasTranslation([NotNull] Func<IReadOnlyCollecti
             return this;
         }
 
+
+        /// <summary>
+        ///     This API supports the Entity Framework Core infrastructure and is not intended to be used 
+        ///     directly from your code. This API may change or be removed in future releases.
+        /// </summary>
+        public virtual DbFunctionParameterBuilder HasParameter([NotNull] string name)
```
Comment (smitpatel) [inline src/EFCore.Relational/Metadata/Builders/DbFunctionBuilder.cs]: @AndriySvyryd - Parameters are already defined looking at methodInfo. Here we want to get particular parameter to configure it. What should be right fluent way to do it? Calling HasParameter with name which is not any parameter in function throws

**Comment by smitpatel** [comment]
This actually does not add any API to configure the parameter & functions return type's database type. Let me discuss with team and get back to you in terms of API shape and data storage necessary.
Columns remember the annotation which specify string format of the type on database. If we don't store that info, I am not sure how would we get type mapping source to use that info to set the type mapping accurately.

File: src/EFCore.Relational/Metadata/Builders/DbFunctionParameterBuilder.cs (hunk 8 above)
Comment (pmiddleton) [reply src/EFCore.Relational/Metadata/Builders/DbFunctionParameterBuilder.cs]: This was leftover from when I originally wrote this a few years ago before I was auto pulling the values from the metadata.  

That being said Oracle does support calling functions with named parameters.  If for some reason the db function parameter name didn't match up with the code name then you could reset it.  That assumes the backend Oracle provider would support this.  

That can be an enhancement down the road and I will remove this.

File: src/EFCore.Relational/Metadata/Internal/DbFunction.cs (hunk 11 above)
Comment (pmiddleton) [reply src/EFCore.Relational/Metadata/Internal/DbFunction.cs]: That's what I thought as well, however Resharper was giving me a "possible null assignment warning".  I added the ? to shut it up.

File: src/EFCore.Relational/Metadata/IMutableDbFunctionParameter.cs (hunk 7 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Metadata/IMutableDbFunctionParameter.cs]: I'll add those later myself

File: src/EFCore.Relational/Metadata/Internal/DbFunction.cs (hunk 11 above)
Comment (smitpatel) [reply src/EFCore.Relational/Metadata/Internal/DbFunction.cs]: Add R# disable instead.

**Comment by ajcvickers** [comment]
When merging, check if #14990 is still needed.

File: src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs (hunk 4 above)
Comment (pmiddleton) [reply src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs]: @smitpatel - Based on what you have said across these comments there won't be anything inside of the Translation == null check.  Is this appropriate for the new query pipeline?  Every parameter should have a type mapping, even if the translation is done manually?

**Comment by pmiddleton** [comment]
@smitpatel  Did you confirm with the team that a return type parameter mapping on dbfunction is required?  If so I will get it added in.

File: src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs (hunk 4 above)
Comment (smitpatel) [reply src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs]: That is interesting question. Likely.

**Comment by smitpatel** [comment]
@pmiddleton - I haven't got a chance yet. Give me about 2 more weeks to sort it out. Presently I am dealing with bigger chunks missing in new pipeline since it is blocking AspNetCore Identity to take new EF packages.

**Comment by pmiddleton** [comment]
@smitpatel - No problem.  I am moving in 2 weeks so I don't have a ton of time right now myself.  We will get this done in June.

**Comment by smitpatel** [comment]
Design meeting decision:
There will be API taking string type name only. Which would map to TypeMapping.

**Comment by smitpatel** [comment]
@pmiddleton - Would you have chance to complete this in next 1 week? If not I can take over and make modifications to merge. Release deadlines to meet.

**Comment by pmiddleton** [comment]
@smitpatel - sorry to do this to you, but I'm not going to be able to get to this in that time frame.  Too much stuff to deal with from my move.  We are still trying to stop living out of boxes :)

**Comment by smitpatel** [comment]
@pmiddleton - No worries. I will wrap up the pending work.

**Comment by smitpatel** [comment]
@pmiddleton - I have created #16182 
Thanks for contribution.

**Comment by pmiddleton** [comment]
@smitpatel  Thanks for finishing this up for me.  Sorry I wasn't able to get this done.
